### PR TITLE
[Program: GCI] create dockerfile for backend

### LINKED ISSUE #266: Create a Dockerfile
## Description

As a developer,
I need a Dockerfile,
so that I can containerise the application and deploy it without having to setup the whole local environment.

The idea is to create a Dockerfile using a python image, probably.
Avoid doing unnecessary setup within the docker image.
Once this works, provide clear instructions on the README file of how to use [Docker](https://www.docker.com/) to deploy the application.

### Update [Required]
- [ ] Create a Dockerfile that allows to deploy the app with Docker
- [ ] Add section to the README explaining how to use Docker to deploy the application

## Definition of Done
- [ ] All of the required items are completed.
- [ ] Approval by 1 mentor.

## Estimation
4 hours

### PR DESCRIPTION
**Pull request by daksha-aeer**
### Description
Create a Dockerfile to deploy the backend without setting up local environment.

#### Environment used
Official python 3.7 image https://hub.docker.com/_/python
```
FROM python:3.7
```

Fixes #266 

### Type of Change:
- Code
- Documentation

**Code/Quality Assurance Only**
- This change requires a documentation update (software upgrade on readme file)
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?
1. Create image
```sh
docker build -t mentorship-backend:latest .
```
image

2. Deploy image
Docker container ports must be mapped to the host machine port using --publish so they're visible.
```sh
docker run --env "FLASK_APP=run.py" --publish 5000:5000 mentorship-backend:latest
```
task27_2

image

3. Run tests
```sh
python -m unittest discover tests
```
task27_3

#### API test on docker
**Create user**
image

**Login**
image

### Checklist:

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have made corresponding changes to the documentation
- [x] Update requirements.txt

**Code/Quality Assurance Only**
- [x] My changes generate no new warnings 
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -34,7 +34,7 @@ PyJWT==1.4.2
 python-dateutil==2.7.3
 python-dotenv==0.8.2
 pytz==2018.4
-PyYAML==3.12
```
Comment (daksha-aeer) [inline requirements.txt:37]: This version was giving error so I updated it. Based on a mentor's advice I updated PyYAML to 3.13 version (as suggested by task https://codein.withgoogle.com/tasks/5320953601458176/?sp-is_beginner=False)

```sh
    Running setup.py install for PyYAML: finished with status 'error'
    ERROR: Command errored out with exit status 1:
     command: /usr/local/bin/python -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-1z0vzype/PyYAML/setup.py'"'"'; __file__='"'"'/tmp/pip-install-1z0vzype/PyYAML/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-44_ibqd3/install-record.txt --single-version-externally-managed --compile
         cwd: /tmp/pip-install-1z0vzype/PyYAML/
    Complete output (496 lines):
    running install
    running build
    running build_py
    creating build
    creating build/lib.linux-x86_64-3.7
    creating build/lib.linux-x86_64-3.7/yaml
    copying lib3/yaml/error.py -> build/lib.linux-x86_64-3.7/yaml
    copying lib3/yaml/parser.py -> build/lib.linux-x86_64-3.7/yaml
    copying lib3/yaml/dumper.py -> build/lib.linux-x86_64-3.7/yaml
    copying lib3/yaml/composer.py -> build/lib.linux-x86_64-3.7/yaml
    copying lib3/yaml/__init__.py -> build/lib.linux-x86_64-3.7/yaml
    copying lib3/yaml/resolver.py -> build/lib.linux-x86_64-3.7/yaml
    copying lib3/yaml/emitter.py -> build/lib.linux-x86_64-3.7/yaml
    copying lib3/yaml/loader.py -> build/lib.linux-x86_64-3.7/yaml
    copying lib3/yaml/events.py -> build/lib.linux-x86_64-3.7/yaml
    copying lib3/yaml/reader.py -> build/lib.linux-x86_64-3.7/yaml
    copying lib3/yaml/cyaml.py -> build/lib.linux-x86_64-3.7/yaml
    copying lib3/yaml/tokens.py -> build/lib.linux-x86_64-3.7/yaml
    copying lib3/yaml/scanner.py -> build/lib.linux-x86_64-3.7/yaml
    copying lib3/yaml/representer.py -> build/lib.linux-x86_64-3.7/yaml
    copying lib3/yaml/serializer.py -> build/lib.linux-x86_64-3.7/yaml
    copying lib3/yaml/constructor.py -> build/lib.linux-x86_64-3.7/yaml
    copying lib3/yaml/nodes.py -> build/lib.linux-x86_64-3.7/yaml
    running build_ext
    creating build/temp.linux-x86_64-3.7
    checking if libyaml is compilable
    gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -I/usr/local/include/python3.7m -c build/temp.linux-x86_64-3.7/check_libyaml.c -o build/temp.linux-x86_64-3.7/check_libyaml.o
    checking if libyaml is linkable
    gcc -pthread build/temp.linux-x86_64-3.7/check_libyaml.o -L/usr/local/lib -lyaml -o build/temp.linux-x86_64-3.7/check_libyaml
    building '_yaml' extension
    creating build/temp.linux-x86_64-3.7/ext
    gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -I/usr/local/include/python3.7m -c ext/_yaml.c -o build/temp.linux-x86_64-3.7/ext/_yaml.o
    In file included from ext/_yaml.c:271:
    ext/_yaml.h:10: warning: "PyString_CheckExact" redefined
     #define PyString_CheckExact PyBytes_CheckExact

    ext/_yaml.c:139: note: this is the location of the previous definition
       #define PyString_CheckExact          PyUnicode_CheckExact

    ext/_yaml.c: In function ‘__pyx_pf_5_yaml_get_version_string’:
    ext/_yaml.c:1410:17: warning: assignment discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
       __pyx_v_value = yaml_get_version_string();
                     ^
    ext/_yaml.c: In function ‘__pyx_pf_5_yaml_7CParser___init__’:
    ext/_yaml.c:2577:52: warning: passing argument 2 of ‘yaml_parser_set_input’ from incompatible pointer type [-Wincompatible-pointer-types]
         yaml_parser_set_input((&__pyx_v_self->parser), __pyx_f_5_yaml_input_handler, ((void *)__pyx_v_self));
                                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
    In file included from ext/_yaml.h:2,
                     from ext/_yaml.c:271:
    /usr/include/yaml.h:1370:30: note: expected ‘int (*)(void *, unsigned char *, size_t,  size_t *)’ {aka ‘int (*)(void *, unsigned char *, long unsigned int,  long unsigned int *)’} but argument is of type ‘int (*)(void *, char *, size_t,  size_t *)’ {aka ‘int (*)(void *, char *, long unsigned int,  long unsigned int *)’}
             yaml_read_handler_t *handler, void *data);
             ~~~~~~~~~~~~~~~~~~~~~^~~~~~~
    In file included from /usr/local/include/python3.7m/Python.h:98,
                     from ext/_yaml.c:4:
    /usr/local/include/python3.7m/bytesobject.h:87:57: warning: pointer targets in passing argument 2 of ‘yaml_parser_set_input_string’ differ in signedness [-Wpointer-sign]
     #define PyBytes_AS_STRING(op) (assert(PyBytes_Check(op)), \
                                   ~~~~~~~~~~~~~~~~~~~~~~~~~~^~~
                                     (((PyBytesObject *)(op))->ob_sval))
                                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    ext/_yaml.h:11:29: note: in expansion of macro ‘PyBytes_AS_STRING’
     #define PyString_AS_STRING  PyBytes_AS_STRING
                                 ^~~~~~~~~~~~~~~~~
    ext/_yaml.c:2818:59: note: in expansion of macro ‘PyString_AS_STRING’
         yaml_parser_set_input_string((&__pyx_v_self->parser), PyString_AS_STRING(__pyx_v_stream), PyString_GET_SIZE(__pyx_v_stream));
                                                               ^~~~~~~~~~~~~~~~~~
    In file included from ext/_yaml.h:2,
                     from ext/_yaml.c:271:
    /usr/include/yaml.h:1343:1: note: expected ‘const unsigned char *’ but argument is of type ‘char *’
     yaml_parser_set_input_string(yaml_parser_t *parser,
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
    ext/_yaml.c: In function ‘__pyx_f_5_yaml_7CParser__token_to_object’:
    ext/_yaml.c:4572:71: warning: pointer targets in passing argument 1 of ‘PyUnicode_FromString’ differ in signedness [-Wpointer-sign]
         __pyx_t_2 = PyUnicode_FromString(__pyx_v_token->data.tag_directive.handle); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 417, __pyx_L1_error)
                                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~
    In file included from /usr/local/include/python3.7m/Python.h:99,
                     from ext/_yaml.c:4:
    /usr/local/include/python3.7m/unicodeobject.h:702:23: note: expected ‘const char *’ but argument is of type ‘yaml_char_t *’ {aka ‘unsigned char *’}
     PyAPI_FUNC(PyObject*) PyUnicode_FromString(
                           ^~~~~~~~~~~~~~~~~~~~
    ext/_yaml.c:4584:71: warning: pointer targets in passing argument 1 of ‘PyUnicode_FromString’ differ in signedness [-Wpointer-sign]
         __pyx_t_2 = PyUnicode_FromString(__pyx_v_token->data.tag_directive.prefix); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 418, __pyx_L1_error)
                                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~
    In file included from /usr/local/include/python3.7m/Python.h:99,
                     from ext/_yaml.c:4:
    /usr/local/include/python3.7m/unicodeobject.h:702:23: note: expected ‘const char *’ but argument is of type ‘yaml_char_t *’ {aka ‘unsigned char *’}
     PyAPI_FUNC(PyObject*) PyUnicode_FromString(
                           ^~~~~~~~~~~~~~~~~~~~
    ext/_yaml.c:5444:63: warning: pointer targets in passing argument 1 of ‘PyUnicode_FromString’ differ in signedness [-Wpointer-sign]
         __pyx_t_2 = PyUnicode_FromString(__pyx_v_token->data.alias.value); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 448, __pyx_L1_error)
                                          ~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~
    In file included from /usr/local/include/python3.7m/Python.h:99,
                     from ext/_yaml.c:4:
    /usr/local/include/python3.7m/unicodeobject.h:702:23: note: expected ‘const char *’ but argument is of type ‘yaml_char_t *’ {aka ‘unsigned char *’}
     PyAPI_FUNC(PyObject*) PyUnicode_FromString(
                           ^~~~~~~~~~~~~~~~~~~~
    ext/_yaml.c:5518:64: warning: pointer targets in passing argument 1 of ‘PyUnicode_FromString’ differ in signedness [-Wpointer-sign]
         __pyx_t_2 = PyUnicode_FromString(__pyx_v_token->data.anchor.value); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 451, __pyx_L1_error)
                                          ~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~
    In file included from /usr/local/include/python3.7m/Python.h:99,
                     from ext/_yaml.c:4:
    /usr/local/include/python3.7m/unicodeobject.h:702:23: note: expected ‘const char *’ but argument is of type ‘yaml_char_t *’ {aka ‘unsigned char *’}
     PyAPI_FUNC(PyObject*) PyUnicode_FromString(
                           ^~~~~~~~~~~~~~~~~~~~
    ext/_yaml.c:5592:61: warning: pointer targets in passing argument 1 of ‘PyUnicode_FromString’ differ in signedness [-Wpointer-sign]
         __pyx_t_2 = PyUnicode_FromString(__pyx_v_token->data.tag.handle); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 454, __pyx_L1_error)
                                          ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~
    In file included from /usr/local/include/python3.7m/Python.h:99,
                     from ext/_yaml.c:4:
    /usr/local/include/python3.7m/unicodeobject.h:702:23: note: expected ‘const char *’ but argument is of type ‘yaml_char_t *’ {aka ‘unsigned char *’}
     PyAPI_FUNC(PyObject*) PyUnicode_FromString(
                           ^~~~~~~~~~~~~~~~~~~~
    ext/_yaml.c:5604:61: warning: pointer targets in passing argument 1 of ‘PyUnicode_FromString’ differ in signedness [-Wpointer-sign]
         __pyx_t_2 = PyUnicode_FromString(__pyx_v_token->data.tag.suffix); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 455, __pyx_L1_error)
                                          ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~
    In file included from /usr/local/include/python3.7m/Python.h:99,
                     from ext/_yaml.c:4:
    /usr/local/include/python3.7m/unicodeobject.h:702:23: note: expected ‘const char *’ but argument is of type ‘yaml_char_t *’ {aka ‘unsigned char *’}
     PyAPI_FUNC(PyObject*) PyUnicode_FromString(
                           ^~~

```

File: run.py [hunk 2]
```diff
@@ -35,6 +35,5 @@ def create_tables():
     from app.database.sqlalchemy_extension import db
     db.create_all()
 
-
 if __name__ == "__main__":
-    application.run(port=5000)
+    application.run(port=5000, host='0.0.0.0')
```
Comment (daksha-aeer) [inline run.py]: Backend won't work in browser without ```host='0.0.0.0'```. Tried with ```host='127.0.0.1'``` but this didn't work.

**Review by sumit-badsara — APPROVED** [review]
It works great! No issues encountered. And as mentioned you have bumped up the versions of PyYAML and used the Python 3.7 image and added the 'how to use' section in Readme. Great! :fire:
Good to go from my side! :smile:

File: run.py (hunk 2 above)
Comment (m-murad) [reply run.py]: are you using the port in address?

Comment (daksha-aeer) [reply run.py]: Yes. The API is visible only on 0.0.0.0:5000. You can see the swagger screenshot. According to this article 
https://pythonspeed.com/articles/docker-connection-refused/ found by @bartek-pacia , this step is necessary

File: README.md [hunk 3]
```diff
@@ -46,6 +46,19 @@ export APP_MAIL_PASSWORD=<app-mail-password>
...
 
+### Run with docker
+1. Make sure you exported the following [environment variables](https://github.com/systers/mentorship-backend/wiki/Environment-Variables) to .env file
+
+2. Build docker image
+```
+docker build -t mentorship-backend:latest .
+```
+3. Deploy
+Docker container ports must be mapped to the host machine port using --publish so they're visible.
```
Comment (isabelcosta) [inline README.md]:
```suggestion
Docker container ports must be mapped to the host machine port using `--publish` so they're visible.
```

File: README.md [hunk 4]
```diff
@@ -46,6 +46,19 @@ export APP_MAIL_PASSWORD=<app-mail-password>
 6. When you are done using the app, deactivate the virtual environment:
 `deactivate`
 
+### Run with docker
+1. Make sure you exported the following [environment variables](https://github.com/systers/mentorship-backend/wiki/Environment-Variables) to .env file
```
Comment (isabelcosta) [inline README.md]:
```suggestion
1. Make sure you exported the following [environment variables](https://github.com/systers/mentorship-backend/wiki/Environment-Variables) to `.env` file
```

**Review by isabelcosta — CHANGES_REQUESTED** [review]
Thank you so much for your contribution with this task @daksha-aeer 🎉 

Have you tried to run this and create a user? Use a feature from the api? I say this because I tried doing this dockerfile myself, and I had to add a timezone to the background scheduler because a feature needed it.

**Comment by daksha-aeer** [comment]
@isabelcosta I could create a user and login using docker image without any problem. I'm using **Ubuntu 19.10**
image

image

@bartekpacia encountered timezone issue as well. I think the docker image should be tested on different operating systems- linux, mac and windows. I'm ready to work on this but unfortunately my Windows 10 Home doesn't allow docker.

File: README.md (hunk 3 above)
Comment (daksha-aeer) [reply README.md]: Done

File: README.md (hunk 4 above)
Comment (daksha-aeer) [reply README.md]: Done

File: dockerfile [hunk 5]
```diff
@@ -0,0 +1,7 @@
+FROM python:3.7
+COPY ./requirements.txt /dockerBuild/requirements.txt
+WORKDIR /dockerBuild
+RUN pip install -r requirements.txt
+COPY . /dockerBuild
+ENTRYPOINT [ "python" ]
+CMD [ "run.py" ]
```
Comment (sumit-badsara) [inline dockerfile]: Better replace this command with: explicit mention of host `0.0.0.0` here itself rather than changing the default in `run.py`

**Review by sumit-badsara — CHANGES_REQUESTED** [review]
Could you add `.dockerignore` and add relevant files to it, also research a bit on which files shouldn't be added to the image of docker as it could increase the size substantially. So, if possible attach the link of reference.

File: dockerfile (hunk 5 above)
Comment (daksha-aeer) [reply dockerfile]: Done!

**Comment by daksha-aeer** [comment]
> Could you add `.dockerignore` and add relevant files to it, also research a bit on which files shouldn't be added to the image of docker as it could increase the size substantially. So, if possible attach the link of reference.

I've added **.dockerignore** file. Here's the link https://codefresh.io/docker-tutorial/not-ignore-dockerignore-2/

**Review by sumit-badsara — APPROVED** [review]

File: README.md [hunk 6]
```diff
@@ -46,6 +46,19 @@ export APP_MAIL_PASSWORD=<app-mail-password>
...
+1. Make sure you exported the following [environment variables](https://github.com/systers/mentorship-backend/wiki/Environment-Variables) to '.env' file
+
+2. Build docker image
+```
+docker build -t mentorship-backend:latest .
+```
+3. Deploy
+Docker container ports must be mapped to the host machine port using '--publish' so they're visible.
+```sh
+docker run --env "FLASK_APP=run.py" --publish 5000:5000 mentorship-backend:latest
```
Comment (isabelcosta) [inline README.md:59]: Do you need this flask app env variable here? isn't this solved with the new CMD on Docker file?

**Review by isabelcosta — APPROVED** [review]
Looks good @daksha-aeer ! thank you for your contribution :)

**Comment by isabelcosta** [comment]
Also this is not a new feature. Could you please, @daksha-aeer could change the commit message or @BadduCoder could change the commit when merging, to start as `chore:` [(our commit message style guideline)](https://github.com/systers/mentorship-backend/wiki/Commit-Message-Style-Guide)

**Comment by sumit-badsara** [comment]
> Also this is not a new feature. Could you please, @daksha-aeer could change the commit message or @BadduCoder could change the commit when merging, to start as `chore:` [(our commit message style guideline)](https://github.com/systers/mentorship-backend/wiki/Commit-Message-Style-Guide)

@daksha-aeer In case you can do this, it would be great! Or I will do once I merge this PR.

**Comment by daksha-aeer** [comment]
Done :)

**Comment by isabelcosta** [comment]
@daksha-aeer you changed the title of the PR, I mentioned the git commit :) @BadduCoder can do that if you prefer ;)

**Comment by daksha-aeer** [comment]
I'm so sorry I got a bit frazzled 😅. I'll change the git commit message in about 20 mins. I'll be reaching home soon.

**Comment by isabelcosta** [comment]
No problem at all! we have time :) and its not super required for you to do it ;)

**Comment by daksha-aeer** [comment]
Done :)

**Pull request merged**